File: scala-jps-plugin/src/org/jetbrains/jps/incremental/scala/data/CompilationData.java
package org.jetbrains.jps.incremental.scala.data;

import org.jetbrains.jps.ModuleChunk;
import org.jetbrains.jps.incremental.CompileContext;
import org.jetbrains.jps.incremental.ModuleBuildTarget;
import org.jetbrains.jps.incremental.scala.ConfigurationException;
import org.jetbrains.jps.incremental.scala.SettingsManager;
import org.jetbrains.jps.incremental.scala.model.CompilerLibraryHolder;
import org.jetbrains.jps.incremental.scala.model.FacetSettings;
import org.jetbrains.jps.incremental.scala.model.LibraryLevel;
import org.jetbrains.jps.incremental.scala.model.ProjectSettings;
import org.jetbrains.jps.model.JpsModel;
import org.jetbrains.jps.model.JpsProject;
import org.jetbrains.jps.model.library.JpsLibrary;
import org.jetbrains.jps.model.library.JpsLibraryCollection;
import org.jetbrains.jps.model.library.JpsOrderRootType;
import org.jetbrains.jps.model.module.JpsModule;

import java.io.File;
import java.util.Collection;

/**
 * @author Pavel Fatin
 */
public class CompilationData {
  private File[] myScalaCompilerClasspath;
  private String[] myScalaCompilerOptions;
  private File myOutputDirectory;
  private File[] myCompilationClasspath;
  private boolean myScalaFirst;

  private CompilationData(File[] scalaCompilerClasspath,
                          String[] scalaCompilerOptions,
                          File outputDirectory,
                          File[] compilationClasspath,
                          boolean scalaFirst) {
    myScalaCompilerClasspath = scalaCompilerClasspath;
    myScalaCompilerOptions = scalaCompilerOptions;
    myOutputDirectory = outputDirectory;
    myCompilationClasspath = compilationClasspath;
    myScalaFirst = scalaFirst;
  }

  public File[] getScalaCompilerClasspath() {
    return myScalaCompilerClasspath;
  }

  public String[] getScalaCompilerOptions() {
    return myScalaCompilerOptions;
  }

  public File getOutputDirectory() {
    return myOutputDirectory;
  }

  public File[] getCompilationClasspath() {
    return myCompilationClasspath;
  }

  public boolean isScalaFirst() {
    return myScalaFirst;
  }

  public static CompilationData create(CompileContext context, ModuleChunk chunk) {
    JpsModule module = chunk.representativeTarget().getModule();
    JpsModel model = context.getProjectDescriptor().getModel();

    // Find a Scala compiler library that is configured in a Scala facet
    JpsLibrary compilerLibrary = getCompilerLibraryIn(module, model);

    // Collect all files in the compiler library
    Collection<File> compilerClasspath = compilerLibrary.getFiles(JpsOrderRootType.COMPILED);
    if (compilerClasspath.isEmpty()) {
      throw new ConfigurationException("Scala compiler library is empty: " + compilerLibrary.getName());
    }

    ModuleBuildTarget target = chunk.representativeTarget();

    // Get an output directory
    File outputDirectory = target.getOutputDir();
    if (outputDirectory == null) {
      throw new ConfigurationException("Output directory not specified for module " + target.getModuleName());
    }

    // Get compilation classpath files
    Collection<File> chunkClasspath = context.getProjectPaths()
        .getCompilationClasspathFiles(chunk, chunk.containsTests(), false, false);

    FacetSettings facet = SettingsManager.getFacetSettings(module);
    String[] compilerOptions = facet.getCompilerOptions();

    ProjectSettings projectSettings = SettingsManager.getProjectSettings(model.getProject());
    boolean scalaFirst = projectSettings.isScalaFirst();

    return new CompilationData(compilerClasspath.toArray(new File[compilerClasspath.size()]), compilerOptions,
        outputDirectory, chunkClasspath.toArray(new File[chunkClasspath.size()]), scalaFirst);
  }

  private static JpsLibrary getCompilerLibraryIn(JpsModule module, JpsModel model) {
    FacetSettings facet = SettingsManager.getFacetSettings(module);

    if (facet == null) {
      throw new ConfigurationException("No Scala facet in module " + module.getName());
    }

    JpsProject project = model.getProject();
    ProjectSettings projectSettings = SettingsManager.getProjectSettings(project);

    boolean fsc = facet.isFscEnabled();

    // Use either a facet compiler library or a project FSC compiler library
    CompilerLibraryHolder libraryHolder = fsc ? projectSettings : facet;

    LibraryLevel compilerLibraryLevel = libraryHolder.getCompilerLibraryLevel();

    if (compilerLibraryLevel == null) {
      String message = fsc
          ? "No FSC compiler library level set in project " + project.getName()
          : "No compiler library level set in module " + module.getName();
      throw new ConfigurationException(message);
    }

    JpsLibraryCollection libraryCollection = getLibraryCollection(compilerLibraryLevel, model, module);

    String compilerLibraryName = libraryHolder.getCompilerLibraryName();

    if (compilerLibraryName == null) {
      String message = fsc
          ? "No FSC compiler library name set in project " + project.getName()
          : "No compiler library name set in module " + module.getName();
      throw new ConfigurationException(message);
    }

    JpsLibrary library = libraryCollection.findLibrary(compilerLibraryName);

    if (library == null) {
      String message = fsc
          ? String.format("FSC compiler library in project %s not found: %s / %s ",
          project.getName(), compilerLibraryLevel, compilerLibraryName)
          : String.format("Сompiler library for module %s not found: %s / %s ",
          module.getName(), compilerLibraryLevel, compilerLibraryName);
      throw new ConfigurationException(message);
    }

    return library;
  }

  private static JpsLibraryCollection getLibraryCollection(LibraryLevel level, JpsModel model, JpsModule module) {
    switch (level) {
      case Global:
        return model.getGlobal().getLibraryCollection();
      case Project:
        return model.getProject().getLibraryCollection();
      case Module:
        return module.getLibraryCollection();
      default:
        throw new ConfigurationException("Unknown library level: " + level);
    }
  }
}

File: jps-plugin/src/org/jetbrains/jps/incremental/scala/data/CompilationData.scala
package org.jetbrains.jps.incremental.scala
package data

import java.io.{File, IOException}
import java.util
import java.util.Collections

import org.jetbrains.jps.builders.java.{JavaBuilderUtil, JavaModuleBuildTargetType}
import org.jetbrains.jps.incremental.java.JavaBuilder
import org.jetbrains.jps.incremental.scala.model.{CompileOrder, CompilerSettings}
import org.jetbrains.jps.incremental.{CompileContext, ModuleBuildTarget}
import org.jetbrains.jps.model.java.JpsJavaExtensionService
import org.jetbrains.jps.model.java.compiler.JpsJavaCompilerOptions
import org.jetbrains.jps.{ModuleChunk, ProjectPaths}

import scala.collection.JavaConverters._

case class ZincData(allSources: Seq[File], compilationStartDate: Long, isCompile: Boolean)

/**
 * @author Pavel Fatin
 */
case class CompilationData(sources: Seq[File],
                           classpath: Seq[File],
                           output: File,
                           scalaOptions: Seq[String],
                           javaOptions: Seq[String],
                           order: CompileOrder,
                           cacheFile: File,
                           outputToCacheMap: Map[File, File],
                           outputGroups: Seq[(File, File)],
                           zincData: ZincData){
  def allSourceFilesCount: Option[Int] = Some(zincData.allSources.size)
}

object CompilationData {
  private val compilationStamp = System.nanoTime()

  def from(sources: Seq[File], allSources: Seq[File], context: CompileContext, chunk: ModuleChunk): Either[String, CompilationData] = {
    val target = chunk.representativeTarget
    val module = target.getModule

    outputsNotSpecified(chunk) match {
      case Some(message) => return Left(message)
      case None =>
    }
    val output = target.getOutputDir.getCanonicalFile
    checkOrCreate(output)

    val classpath = ProjectPaths.getCompilationClasspathFiles(chunk, chunk.containsTests, false, true).asScala.toSeq
    val compilerSettings = SettingsManager.getProjectSettings(module.getProject).getCompilerSettings(chunk)
    val scalaOptions = scalaOptionsFor(compilerSettings, chunk)
    val order = compilerSettings.getCompileOrder

    createOutputToCacheMap(context).map { outputToCacheMap =>

      val cacheFile = outputToCacheMap.getOrElse(output, {
        val message =
          s"""Unknown build target output directory: $output
             |Current outputs:
             |${outputToCacheMap.keys.mkString("\n")}
           """.stripMargin
        throw new RuntimeException(message)
      })

      val relevantOutputToCacheMap = (outputToCacheMap - output).filter(p => classpath.contains(p._1))

      val commonOptions = {
        val encoding = context.getProjectDescriptor.getEncodingConfiguration.getPreferredModuleChunkEncoding(chunk)
        Option(encoding).map(Seq("-encoding", _)).getOrElse(Seq.empty)
      }

      val javaOptions = javaOptionsFor(context, chunk)

      val outputGroups = createOutputGroups(chunk)

      val canonicalSources = sources.map(_.getCanonicalFile)

      val hydraSettings = SettingsManager.getHydraSettings(context.getProjectDescriptor.getProject)
      val scalaVersion = CompilerData.compilerVersion(module)
      val hydraOptions =
        if (hydraSettings.isHydraEnabled && scalaVersion.nonEmpty && hydraSettings.getArtifactPaths.containsKey(scalaVersion.get))
          Seq("-sourcepath", outputGroups.map(_._1).mkString(File.pathSeparator), "-cpus", "2")
        else
          Seq.empty

      val isCompile =
        !JavaBuilderUtil.isCompileJavaIncrementally(context) &&
          !JavaBuilderUtil.isForcedRecompilationAllJavaModules(context)

      CompilationData(canonicalSources, classpath, output, commonOptions ++ scalaOptions ++ hydraOptions, commonOptions ++ javaOptions,
        order, cacheFile, relevantOutputToCacheMap, outputGroups,
        ZincData(allSources, compilationStamp, isCompile))
    }
  }

  def checkOrCreate(output: File) {
    if (!output.exists()) {
      try {
        if (!output.mkdirs()) throw new IOException("Cannot create output directory: " + output.toString)
      } catch {
        case t: Throwable => throw new IOException("Cannot create output directory: " + output.toString, t)
      }
    }
  }

  def outputsNotSpecified(chunk: ModuleChunk): Option[String] = {
    val moduleNames = chunk.getTargets.asScala.filter(_.getOutputDir == null).map(_.getModule.getName)
    moduleNames.toSeq match {
      case Seq() => None
      case Seq(name) => Some(s"Output directory not specified for module $name")
      case names => Some(names.mkString(s"Output directory not specified for modules ", ", ", ""))
    }
  }

  def scalaOptionsFor(compilerSettings: CompilerSettings, chunk: ModuleChunk): Array[String] = {
    val noBootCp = if (CompilerData.needNoBootCp(chunk)) Nil else Seq("-nobootcp", "-javabootclasspath", File.pathSeparator)
    val scalaOptions = noBootCp ++: compilerSettings.getCompilerOptions
    scalaOptions
  }

  def javaOptionsFor(context: CompileContext, chunk: ModuleChunk): Seq[String] = {
    val compilerConfig = {
      val project = context.getProjectDescriptor.getProject
      JpsJavaExtensionService.getInstance.getOrCreateCompilerConfiguration(project)
    }

    val options = new util.ArrayList[String]()

    addCommonJavacOptions(options, compilerConfig.getCurrentCompilerOptions)

    val annotationProcessingProfile = {
      val module = chunk.representativeTarget.getModule
      compilerConfig.getAnnotationProcessingProfile(module)
    }

    JavaBuilder.addCompilationOptions(options, context, chunk, annotationProcessingProfile)

    options.asScala
  }

  // TODO JavaBuilder.loadCommonJavacOptions should be public
  def addCommonJavacOptions(options: util.ArrayList[String], compilerOptions: JpsJavaCompilerOptions) {
    if (compilerOptions.DEBUGGING_INFO) {
      options.add("-g")
    }

    if (compilerOptions.DEPRECATION) {
      options.add("-deprecation")
    }

    if (compilerOptions.GENERATE_NO_WARNINGS) {
      options.add("-nowarn")
    }

    if (!compilerOptions.ADDITIONAL_OPTIONS_STRING.isEmpty) {
      // TODO extract VM options
      options.addAll(compilerOptions.ADDITIONAL_OPTIONS_STRING.split("\\s+").toSeq.asJava)
    }
  }

  private def createOutputToCacheMap(context: CompileContext): Either[String, Map[File, File]] = {
    val targetToOutput = targetsIn(context).collect {
      case target if target.getOutputDir != null => (target, target.getOutputDir)
    }

    outputClashesIn(targetToOutput).toLeft {
      val paths = context.getProjectDescriptor.dataManager.getDataPaths

      for ((target, output) <- targetToOutput.toMap)
      yield (
        output.getCanonicalFile,
        new File(
          paths.getTargetDataRoot(target).getCanonicalFile,
          s"cache-${target.getPresentableName}.zip")
      )
    }
  }

  private def createOutputGroups(chunk: ModuleChunk): Seq[(File, File)] = {
    for {
      target <- chunk.getTargets.asScala.toSeq
      if target.getOutputDir != null
      module = target.getModule
      output = target.getOutputDir.getCanonicalFile
      sourceRoot <- module.getSourceRoots.asScala.map(_.getFile.getCanonicalFile)
      if sourceRoot.exists
    } yield (sourceRoot, output)
  }

  private def targetsIn(context: CompileContext): Seq[ModuleBuildTarget] = {
    def isExcluded(target: ModuleBuildTarget): Boolean = {
      val chunk = new ModuleChunk(Collections.singleton(target))
      ChunkExclusionService.isExcluded(chunk)
    }

    def isProductionTargetOfTestModule(target: ModuleBuildTarget): Boolean = {
      target.getTargetType == JavaModuleBuildTargetType.PRODUCTION &&
        JpsJavaExtensionService.getInstance.getTestModuleProperties(target.getModule) != null
    }

    val buildTargetIndex = context.getProjectDescriptor.getBuildTargetIndex
    val targets = JavaModuleBuildTargetType.ALL_TYPES.asScala.flatMap(buildTargetIndex.getAllTargets(_).asScala)

    targets.distinct.filterNot { target =>
      buildTargetIndex.isDummy(target) || isExcluded(target) || isProductionTargetOfTestModule(target)
    }
  }

  private def outputClashesIn(targetToOutput: Seq[(ModuleBuildTarget, File)]): Option[String] = {
    val outputToTargetsMap = targetToOutput.groupBy(_._2).mapValues(_.map(_._1))

    val errors = outputToTargetsMap.collect {
      case (output, targets) if output != null && targets.length > 1 =>
        val targetNames = targets.map(_.getPresentableName).mkString(", ")
        "Output path %s is shared between: %s".format(output, targetNames)
    }

    if (errors.isEmpty) None else Some(errors.mkString("\n") +
            "\nPlease configure separate output paths to proceed with the compilation." +
            "\nTIP: you can use Project Artifacts to combine compiled classes if needed.")
  }
}